### PR TITLE
feat: add RLock for GRPCClient

### LINKED ISSUE #5527: Thread lock the Jina client (or streamer) using RLock
The Jina GRPCClient cannot be used in multiple threads (due to a limitation from grpc: https://github.com/grpc/grpc/issues/25364).

Use [RLock](https://docs.python.org/3/library/threading.html#rlock-objects) to restrict the usage of grpc client within 1 thread at a time and test this in a multi threaded setup (flask).
You can get inspired by this commit https://github.com/jina-ai/jina/commit/af988e9920b6ff35fb3014f32b9b27d86797bd3b

### PR DESCRIPTION
**Pull request by AnneYang720**
**Goals:**

- resolves #5527

You can try with the scripts below to see if there are `BlockingIOError`. 

flow.py
```python
import os
from jina import Flow
from jina import Executor, requests

os.environ['JINA_LOG_LEVEL'] = 'DEBUG'

class MyExecutor(Executor):
    @requests
    def foo(self, docs, **kwargs):
        for doc in docs:
            doc.text = 'I am coming from MyExecutor'


f = Flow(port=12345).add(uses=MyExecutor, reload=True)

with f:
    f.block()
```

client.py
```python
from jina import Client, Document
from concurrent.futures import ThreadPoolExecutor
from threading import RLock

c = Client(port=12345)

def client_post():
    doc = Document(text='hello world')
    result = c.post(on='/', inputs=doc)[0]
    print(result.text)


with ThreadPoolExecutor(max_workers=50) as t:
    tasks = []
    for _ in range(10000):
        task = t.submit(client_post)
        tasks.append(task)
    
    for task in tasks:
        task.result()
```

### CODE REVIEW

File: tests/integration/concurrent_clients/test_multithread_client.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            doc.text = 'I am coming from MyExecutor'
+                
+def flow_run(flow, stop_event):
+    with flow:
+        flow.block(stop_event)
+
+
+def client_post(c):
+    doc = Document(id='doc', text='hello world')
+    result = c.post(on='/', inputs=doc)[0]
```
Comment (alaeddine-13) [inline tests/integration/concurrent_clients/test_multithread_client.py:22]:
```suggestion
def client_post(client):
    doc = Document(id='doc', text='hello world')
    result = client.post(on='/', inputs=doc)[0]
```

File: tests/integration/concurrent_clients/test_multithread_client.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+        with ThreadPoolExecutor(max_workers=50) as pool:
+            tasks = []
+            for _ in range(1000):
+                task = pool.submit(client_post, c)
+                tasks.append(task)
+
+            for task in tasks:
+                result = task.result()
+                assert result.id == 'doc'
```
Comment (alaeddine-13) [inline tests/integration/concurrent_clients/test_multithread_client.py:46]: can we put unique ids per doc and assert them ?

File: jina/clients/base/grpc.py [hunk 3]
```diff
@@ -203,6 +205,7 @@ async def _get_results(
                 metadata = metadata or ()
                 metadata = metadata + (('__results_in_order__', 'true'),)
 
+            self._lock.acquire()
```
Comment (alaeddine-13) [inline jina/clients/base/grpc.py:208]: can we use with self._lock instead ?

**Review by alaeddine-13 — DISMISSED** [review]

**Comment by alaeddine-13** [comment]
can we update this page ?
https://docs.jina.ai/concepts/client/#connect
it mentions that the client cannot be used in multithreaded environment

**Comment by alaeddine-13** [comment]
Also can we confirm that multi threaded flask server can now use jina client without issues ?
feel free to reach out to @ZiniuYu  about the example:
https://jinaai.slack.com/archives/C018F60RBL5/p1663661829374689

**Comment by alaeddine-13** [comment]
@AnneYang720 can you also experiment with this example:
```python
from jina import Client, Document
from concurrent.futures import ThreadPoolExecutor
from threading import RLock



def client_post():
    c = Client(port=12345)
    doc = Document(text='hello world')
    result = c.post(on='/', inputs=doc)[0]
    print(result.text)


with ThreadPoolExecutor(max_workers=50) as t:
    tasks = []
    for _ in range(10000):
        task = t.submit(client_post)
        tasks.append(task)
    
    for task in tasks:
        task.result()
```

If feel that the lock should be a global object, not created per client object

File: docs/concepts/client/index.md [hunk 4]
```diff
@@ -148,16 +148,6 @@ Client(host='https://my.awesome.flow:1234', port=4321)
 ```
 ````
 
-````{admonition} Caution
-:class: caution
```
Comment (JoanFM) [inline docs/concepts/client/index.md:152]: This should be kept, just mention that we use an RLock so that the effect they expect may not be observed

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/integration/concurrent_clients/test_multithread_client.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+def test_multithread_client(capsys):
+    stop_event = threading.Event()
+    flow = Flow(port=12345).add(uses=MyExecutor)
+    t = threading.Thread(target=flow_run, args=(flow, stop_event))
+
+    c = Client(port=12345)
+
+    try:
+        with capsys.disabled():
```
Comment (JoanFM) [inline tests/integration/concurrent_clients/test_multithread_client.py:33]: what does this do?

File: tests/integration/concurrent_clients/test_multithread_client.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def client_post(doc, client):
+    result = client.post(on='/', inputs=doc)[0]
+    return result
+
+
+def test_multithread_client(capsys):
+    stop_event = threading.Event()
+    flow = Flow(port=12345).add(uses=MyExecutor)
```
Comment (JoanFM) [inline tests/integration/concurrent_clients/test_multithread_client.py:27]: do not use a predefined port, use `random_port`

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by JoanFM** [comment]
> Also can we confirm that multi threaded flask server can now use jina client without issues ? feel free to reach out to @ZiniuYu about the example: https://jinaai.slack.com/archives/C018F60RBL5/p1663661829374689

Also, make sure we can apply the same to `streamer` @alaeddine-13 @AnneYang720

File: tests/integration/concurrent_clients/test_multithread_client.py (hunk 5 above)
Comment (AnneYang720) [reply tests/integration/concurrent_clients/test_multithread_client.py:33]: `capsys` will capture all the `stdout` and `stderr` from the test

File: tests/unit/yaml/dummy_gateway.py [hunk 7]
```diff
@@ -63,3 +75,44 @@ async def run_server(self):
...
+                    docs=docs,
+                    exec_endpoint='/',
+            ):
+                # step 4: convert response docs to server response and return it
+                result = response_docs[0].text
+
+            return {'result': result}
+
+        # step 5: implement health-check
+        @app.route("/", methods=['GET'])
```
Comment (JoanFM) [inline tests/unit/yaml/dummy_gateway.py:102]: please use single quotes

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/yaml/dummy_gateway.py [hunk 8]
```diff
@@ -63,3 +75,44 @@ async def run_server(self):
     async def shutdown(self):
         self.server.should_exit = True
         await self.server.shutdown()
+
+
+class FlaskDummyGateway(Gateway):
+    async def setup_server(self):
+        # step 1: create an app and define the service endpoint
+        app = Flask(__name__)
```
Comment (JoanFM) [inline tests/unit/yaml/dummy_gateway.py:83]: also we need to check how to properly terminate this server (look at `should_exit`)

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: docs/concepts/client/index.md [hunk 9]
```diff
@@ -150,15 +150,15 @@ Client(host='https://my.awesome.flow:1234', port=4321)
 
 ````{admonition} Caution
 :class: caution
-In case you instanciate a `Client` object using the `grpc` protocol, keep in mind that `grpc` clients cannot be used in 
-a multi-threaded environment (check [this gRPC issue](https://github.com/grpc/grpc/issues/25364) for reference).
+
+According to [this gRPC issue](https://github.com/grpc/grpc/issues/25364), `grpc` clients cannot be used in 
+a multi-threaded environment. So we apply `RLock` as a workaround.
```
Comment (JoanFM) [inline docs/concepts/client/index.md:155]: do not call it a workaround

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: docs/concepts/client/index.md (hunk 9 above)
Comment (alexcg1) [reply docs/concepts/client/index.md:155]: Other than what @JoanFM said, I don't see any problem here. Fix the "workaround" thing and ask me for a re-review :wink:

**Review by alexcg1 — APPROVED** [review]
LGTM :+1:

**Pull request merged**